File: AkkaStreamsWithPersistence/src/main/scala/com/lightbend/scala/modelserver/actor/actors/ModelServingActor.scala
package com.lightbend.scala.modelserver.actor.actors

import akka.actor.{Actor, Props}
import com.lightbend.model.winerecord.WineRecord
import com.lightbend.modelServer.model.{Model, ModelToServeStats, ModelWithDescriptor}
import com.lightbend.modelserver.actor.persistence.FilePersistence

// Workhorse - doing model serving for a given data type

class ModelServingActor(dataType : String) extends Actor {

  println(s"Creating model serving actor $dataType")
  private var currentModel: Option[Model] = None
  private var newModel: Option[Model] = None
  var currentState: Option[ModelToServeStats] = None
  private var newState: Option[ModelToServeStats] = None

  override def preStart {
    val state = FilePersistence.restoreState(dataType)
    newState = state._2
    newModel = state._1
  }

  override def receive = {
    case model : ModelWithDescriptor => {
      // Update model
      newState = Some(new ModelToServeStats(model.descriptor))
      newModel = Some(model.model)
      FilePersistence.saveState(dataType, newModel.get, newState.get)
      sender() ! "Done"
    }
    case record : WineRecord => {
      // Process data
      newModel match {
        // Update model
        case Some(model) => {
          // close current model first
          currentModel match {
            case Some(m) => m.cleanup()
            case _ =>
          }
          // Update model
          currentModel = newModel
          currentState = newState
          newModel = None
        }
        case _ =>
      }
      currentModel match {
        case Some(model) => {
          val start = System.currentTimeMillis()
          val quality = model.score(record.asInstanceOf[AnyVal]).asInstanceOf[Double]
          val duration = System.currentTimeMillis() - start
          println(s"Calculated quality - $quality calculated in $duration ms")
          currentState.get.incrementUsage(duration)
          sender() ! Some(quality)
        }
        case _ => {
          println("No model available - skipping")
          sender() ! None
        }
      }
    }
    case request : GetState => {
      // State query
      sender() ! currentState.getOrElse(ModelToServeStats())
    }
  }
}

object ModelServingActor{
  def props(dataType : String) : Props = Props(new ModelServingActor(dataType))
}

case class GetState(dataType : String)

File: akkaStreamsModelServer/src/main/java/com/lightbend/java/akkastreams/modelServer/actors/ModelServingActor.java
package com.lightbend.java.akkastreams.modelServer.actors;

import akka.actor.AbstractActor;
import akka.actor.Props;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import com.lightbend.java.akkastreams.modelServer.actors.presistence.ExecutionState;
import com.lightbend.java.akkastreams.modelServer.actors.presistence.FilePersistence;
import com.lightbend.java.model.Model;
import com.lightbend.java.model.ModelServingInfo;
import com.lightbend.java.model.ModelWithDescriptor;
import com.lightbend.java.model.ServingResult;
import com.lightbend.model.Winerecord;

import java.util.Optional;

public class ModelServingActor extends AbstractActor {
    private final LoggingAdapter log = Logging.getLogger(getContext().getSystem(), this);

    private Optional<Model> currentModel = Optional.empty();
    private Optional<Model> newModel = Optional.empty();
    private Optional<ModelServingInfo> currentServingInfo = Optional.empty();
    private Optional<ModelServingInfo> newServingInfo = Optional.empty();
    private String dataType = null;

    public ModelServingActor(String dataType){
        this.dataType = dataType;
    }

    @Override
    public void preStart() {
        ExecutionState state = FilePersistence.restoreState(dataType);
        newModel = state.getModel();
        newServingInfo = state.getServingInfo();
    }

    private void processModel(ModelWithDescriptor modelWithDescriptor) {

        System.out.println("Processing new model " + modelWithDescriptor);
        newModel = Optional.of(modelWithDescriptor.getModel());
        newServingInfo = Optional.of(new ModelServingInfo(modelWithDescriptor.getDescriptor().getName(),
                modelWithDescriptor.getDescriptor().getDescription(), System.currentTimeMillis()));
        FilePersistence.saveState(dataType, newModel.get(), newServingInfo.get());
        return;
    }

    public ServingResult processData(Winerecord.WineRecord dataRecord) {
        if(newModel.isPresent()){
            // update the model
            if(currentModel.isPresent())
                currentModel.get().cleanup();
            currentModel = newModel;
            currentServingInfo = newServingInfo;
            newModel = Optional.empty();
        }

        // Actually score
        if(currentModel.isPresent()) {
            // Score the model
            long start = System.currentTimeMillis();
            double quality = (double) currentModel.get().score(dataRecord);
            long duration = System.currentTimeMillis() - start;
            currentServingInfo.get().update(duration);
//            System.out.println("Calculated quality - " + quality + " in " + duration + "ms");
            return new ServingResult(quality, duration);
        }
        else{
            // No model currently
//            System.out.println("No model available - skipping");
            return ServingResult.noModel;
        }
    }

    public ModelServingInfo getServingInfo() {
        System.out.println("Using model server " + dataType);
        return currentServingInfo.orElse(ModelServingInfo.empty);
    }

    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(ModelWithDescriptor.class, modelWithDescriptor -> {
                    processModel(modelWithDescriptor);
                    getSender().tell("Done", getSelf());
                })
                .match(Winerecord.WineRecord.class, dataRecord -> {
                    getSender().tell(processData(dataRecord), getSelf());
                })
                .match(GetState.class, dataType -> {
                    getSender().tell(getServingInfo(), getSelf());
                })
                .build();
    }

    public static Props props(String dataType) {
         return Props.create(ModelServingActor.class, () -> new ModelServingActor(dataType));
    }
}
